# LET'S VIEW THE DATA FIRST
The first thing we'll need to do is load in the libraries and datasets we'll be using. For today, I'll be using a dataset of various place of publication of books accross the globe(world)


#     READING THE DATA

In [69]:
# modules we'll use
import pandas as pd
import numpy as np


# set seed for reproducibility
np.random.seed(0)

In [70]:
df = pd.read_csv(r"C:\Users\User\Desktop\Data analysis s\LATEST\python-data-cleaning-master\Datasets\BL-Flickr-Images-Book.csv")

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8287 entries, 0 to 8286
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Identifier              8287 non-null   int64  
 1   Edition Statement       773 non-null    object 
 2   Place of Publication    8287 non-null   object 
 3   Date of Publication     8106 non-null   object 
 4   Publisher               4092 non-null   object 
 5   Title                   8287 non-null   object 
 6   Author                  6509 non-null   object 
 7   Contributors            8287 non-null   object 
 8   Corporate Author        0 non-null      float64
 9   Corporate Contributors  0 non-null      float64
 10  Former owner            1 non-null      object 
 11  Engraver                0 non-null      float64
 12  Issuance type           8287 non-null   object 
 13  Flickr URL              8287 non-null   object 
 14  Shelfmarks              8287 non-null   

The first thing I do when I get a new dataset is take a look at some of it. This lets me see that it all read in correctly and get an idea of what's going on with the data. In this case, I'm looking to see if I see any missing values, which will be reprsented with NaN or None.

In [72]:
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

Looking at the first 5 columns,it is seen that there are lots of missing values in the Data which are automatically represented with NaN

# See how many missing data points we have
Ok, now we know that we do have some missing values. Let's see how many we have in each column.

In [74]:
# get the number of missing data points per column
missing_values_count = df.isnull().sum()

# look at the # of missing points in the first ten columns
missing_values_count[0:10]

Identifier                   0
Edition Statement         7514
Place of Publication         0
Date of Publication        181
Publisher                 4195
Title                        0
Author                    1778
Contributors                 0
Corporate Author          8287
Corporate Contributors    8287
dtype: int64

That seems like a lot! It might be helpful to see what percentage of the values in our dataset were missing to give us a better sense of the scale of this problem:

In [140]:
# how many total missing values do we have?
total_cells = np.product(df.shape)
total_missing = missing_values_count.sum()

#total_cells = total row * total column

print(total_cells)
print(total_missing)

# percent of data that is missing
(total_missing/total_cells) * 100


49722
46815


94.1534934234343

Damn, 94.2% of the cells in this dataset are empty.

#   Dropping unnecessary rows

In [110]:
# remove all the rows that contain a missing value
df.dropna()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London               1879    
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   
...                              ...                 ...   
4156359                     New York                1898   
4157746                       London                1811   
4157862          Newcastle-upon-Tyne                1867   
4158128                        Derby                1831   
4159587          Newcastle upon Tyne                1834   

                         Publisher  \
Identifier                           
206               S. Tinsley & Co.   
216                   Virtue & Co.   
218          Bradbury, Evans & Co.   
472                  James Darling   
480           Wertheim & Macintosh   
...                            ...   
4156359        G. P. Putnam's Sons   
4157746     F. C. and J. Rivington   
4157862                 T. Fordyce   
4158128            M. Mozley & Son   
4159587           Mackenzie & Dent   

                                                        Title  \
Identifier                                                      
206                                             Walter Forbes   
216                                             All For Greed   
218                                          Love The Avenger   
472         Welsh Sketches, Chiefly Ecclesiastical, To The...   
480             The World In Which I Live, And My Place In It   
...                                                       ...   
4156359     History Of The People Of The Netherlands ... T...   
4157746     The New Chronicles Of England And France, In T...   
4157862     Local Records; Or, Historical Register Of Rema...   
4158128          The History And Gazetteer Of The County Of D   
4159587     An Historical, Topographical And Descriptive V...   

                                                   Author  \
Identifier                                                  
206                                                 A. A.   
216                                             A., A. A.   
218                                             A., A. A.   
472                                             A., E. S.   
480                                             A., E. S.   
...                                                   ...   
4156359                               BLOK, Pieter Johan.   
4157746                                   FABYAN, Robert.   
4157862     FORDYCE, T. - Printer, of Newcastle-upon-Tyne   
4158128                        GLOVER, Stephen - of Derby   
4159587                             Mackenzie, E. (Eneas)   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...  
...                                                       ...  
4156359     http://www.flickr.com/photos/britishlibrary/ta...  
4157746     http://www.flickr.com/photos/britishlibrary/ta...  
4157862     http://www.flickr.com/photos/britishlibrary/ta...  
4158128     http://www.flickr.com/photos/britishlibrary/ta...  
4159587     http://www.flickr.com/photos/britishlibrary/ta...  

[2694 rows x 6 columns]

In [116]:
df['Date of Publication'].head(25)

Identifier
206     1879 
216      1868
218      1869
472      1851
480      1857
481      1875
519      1872
667       NaN
874      1676
1143     1679
1280     1802
1808     1859
1905     1888
1929     1839
2836     1897
2854     1865
2956     1860
2957     1873
3017     1866
3131     1899
4598     1814
4884     1820
4976     1800
5382     1847
5385      NaN
Name: Date of Publication, dtype: object

In [139]:
# remove all columns with at least one missing value
columns_with_na_dropped = df.dropna(axis=1)
columns_with_na_dropped.head()

Place of Publication  \
Identifier                             
206                           London   
216         London; Virtue & Yorston   
218                           London   
472                           London   
480                           London   

                                                        Title  \
Identifier                                                      
206                                             Walter Forbes   
216                                             All For Greed   
218                                          Love The Avenger   
472         Welsh Sketches, Chiefly Ecclesiastical, To The...   
480             The World In Which I Live, And My Place In It   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

In [141]:
# just how much data did we lose?
print("Columns in original dataset: %d \n" % df.shape[1])
print("Columns with na's dropped: %d" % columns_with_na_dropped.shape[1])

Columns in original dataset: 6 

Columns with na's dropped: 3


 At this point we have successfully removed all the NaN's from our data.

In [142]:
print("df before dropping all na:", df.shape)
print("df after dropping all na:", df_dropna.shape)


df before dropping all na: (8287, 6)
df after dropping all na: (3062, 6)


The given result before and after dropping na shows that the column has been successfully dropped and we lost quite a large number of data in the row (5225 were missing)

# Cleaning columns using the .apply function

In [146]:
unwanted_characters = ['[', ',', '-']

def clean_dates(item):
    dop= str(item.loc['Date of Publication'])
    
    if dop == 'nan' or dop[0] == '[':
        return np.NaN
    
    for character in unwanted_characters:
        if character in dop:
            character_index = dop.find(character)
            dop = dop[:character_index]
    
    return dop

df['Date of Publication'] = df.apply(clean_dates, axis = 1)
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London               1879    
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                                             Walter Forbes      A. A.   
216                                             All For Greed  A., A. A.   
218                                          Love The Avenger  A., A. A.   
472         Welsh Sketches, Chiefly Ecclesiastical, To The...  A., E. S.   
480             The World In Which I Live, And My Place In It  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...